File: ruby/lib/pay_kit/signer.rb
# frozen_string_literal: true

require "json"

require "pay_core/solana/base58"

require_relative "errors"

module PayKit
  # Factory module for local Ed25519 signers. Every factory returns an
  # object that satisfies the PayKit signer duck-type contract:
  #
  #   #pubkey      → base58 String (44 chars)
  #   #sign(msg)   → 64-byte signature String
  #   #fee_payer?  → Boolean (true for in-process local signers)
  #   #demo?       → Boolean (only true for `Signer.demo`)
  #
  # Remote enclave signers (GCP KMS, AWS KMS, HashiCorp Vault) are
  # reserved under `PayKit::Kms` but are not part of this release; the
  # `Signer::InvalidKeyError` and the contract live here so callers can
  # treat both halves uniformly when the remote backends ship.
  module Signer
    # Raised when an input value cannot be parsed as a valid 64-byte
    # Solana keypair (wrong length, invalid encoding, missing bytes).
    class InvalidKeyError < ::PayKit::Error; end

    module_function

    # The package-shipped demo keypair. Returns the cached `Signer::Demo`
    # instance and emits a one-time `Logger.warn`. Boot-time mainnet
    # refusal is wired in `PayKit::Config#freeze!`.
    def demo
      Demo.instance
    end

    # 64-byte secret as a Ruby Array of integers (Solana CLI keypair
    # format minus the JSON wrapping).
    def bytes(array)
      Local.new(array)
    end

    # Solana CLI JSON-array format, e.g. `"[1,2,3,...,64]"`.
    def json(string)
      array = JSON.parse(string)
      raise InvalidKeyError, "Solana CLI keypair must be a JSON array" unless array.is_a?(Array)

      Local.new(array.map { |element| Integer(element) })
    rescue JSON::ParserError, TypeError => error
      raise InvalidKeyError, "malformed Solana CLI JSON-array keypair: #{error.message}"
    end

    # Phantom / Solflare base58 export form. Solana keypair bytes (64)
    # encoded as base58 produce ~87-88 characters.
    def base58(string)
      decoded = ::PayCore::Solana::Base58.decode(string)
      Local.new(decoded.bytes)
    rescue ArgumentError => error
      raise InvalidKeyError, "malformed base58 keypair: #{error.message}"
    end

    # 128-char hex string (64 bytes hex-encoded).
    def hex(string)
      unless string.is_a?(String) && string.match?(/\A[0-9a-fA-F]+\z/) && string.length.even?
        raise InvalidKeyError, "hex keypair must be an even-length string of hex digits"
      end

      Local.new([string].pack("H*").bytes)
    end

    # Read a Solana CLI JSON-array keypair file.
    def file(path)
      raw = File.read(path)
      json(raw)
    rescue Errno::ENOENT, Errno::EACCES => error
      raise InvalidKeyError, "keypair file unreadable: #{error.message}"
    end

    # Env-var loader. Returns `nil` when the variable is unset or empty
    # so that the caller's default keypair (typically `Signer.demo`)
    # survives the no-op assignment chain. Raises `InvalidKeyError` when
    # the variable holds a value that cannot be parsed in any of the
    # supported formats.
    def env(name)
      raw = ENV[name]
      return nil if raw.nil? || raw.empty?

      stripped = raw.strip
      if stripped.start_with?("[")
        json(stripped)
      elsif stripped.match?(/\A[0-9a-fA-F]{128}\z/)
        hex(stripped)
      else
        base58(stripped)
      end
    end

    # Generate a fresh ephemeral keypair. Test-only utility; production
    # callers should bind to a persistent key source (file/env/KMS).
    def generate
      require "ed25519"

      signing_key = ::Ed25519::SigningKey.generate
      Local.new(signing_key.to_bytes.bytes + signing_key.verify_key.to_bytes.bytes)
    end
  end
end

require_relative "signer/local"
require_relative "signer/demo"

File: lua/pay_kit/signer.lua
--[[
Signer factory family.

Mirrors `PayKit::Signer` in the Ruby SDK and the design's "Signer
backends" section. Local signers (synchronous, no I/O on :sign) ship
in v1; remote KMS-backed signers (cosocket I/O on :sign) live under
`pay_kit.kms` and are reserved for post-v1.

Every factory returns a signer table satisfying the duck type:

  {
    pubkey    = function(self) return "<base58 string>" end,
    sign      = function(self, message) return "<64-byte signature>", err end,
    fee_payer = function(self) return true end,
    demo      = function(self) return false end,
  }

All factories return `(signer, nil)` on success or `(nil, err)` on
malformed input. `from_env` is the one exception: unset / empty env
returns plain `nil` with no error so option-table composition stays
clean.
]]

local local_signer = require('pay_kit.signer.local')
local demo_signer = require('pay_kit.signer.demo')

local M = {}

-- --- helpers ---------------------------------------------------------

local function bytes_table_to_string(t)
  if type(t) ~= 'table' then
    return nil, 'pay_kit: secret must be a table of 64 byte integers'
  end
  if #t ~= 64 then
    return nil, 'pay_kit: secret must be a 64-element table'
  end
  local chars = {}
  for i = 1, 64 do
    local b = t[i]
    if type(b) ~= 'number' or b < 0 or b > 255 or b ~= math.floor(b) then
      return nil, 'pay_kit: secret bytes must be 0..255 integers (index ' .. i .. ')'
    end
    chars[i] = string.char(b)
  end
  return table.concat(chars)
end

local function hex_to_bytes(s)
  if type(s) ~= 'string' or #s ~= 128 then
    return nil, 'pay_kit: hex secret must be a 128-character string'
  end
  if not s:match('^[0-9a-fA-F]+$') then
    return nil, 'pay_kit: hex secret has non-hex characters'
  end
  local out = {}
  for i = 1, 128, 2 do
    out[#out + 1] = string.char(tonumber(s:sub(i, i + 1), 16))
  end
  return table.concat(out)
end

-- --- factories -------------------------------------------------------

-- Hard-coded demo keypair (singleton). Same identity across the
-- Lua and Ruby SDKs.
function M.demo()
  return demo_signer.instance()
end

-- Build a Local signer from a 64-element table of byte integers
-- (Solana CLI keypair JSON's parsed form).
function M.bytes(table_of_ints)
  local str, err = bytes_table_to_string(table_of_ints)
  if err then return nil, err end
  return local_signer.new(str)
end

-- Build a Local signer from a Solana-CLI JSON-array string
-- (`"[1, 2, ..., 64]"`).
function M.json(json_str)
  if type(json_str) ~= 'string' then
    return nil, 'pay_kit: signer.json: expected a string'
  end
  local trimmed = json_str:match('^%s*(.-)%s*$')
  if trimmed == '' then
    return nil, 'pay_kit: signer.json: empty input'
  end
  local cjson_safe = require('cjson.safe')
  local parsed, decode_err = cjson_safe.decode(trimmed)
  if not parsed then
    return nil, 'pay_kit: signer.json: invalid JSON: ' .. tostring(decode_err)
  end
  return M.bytes(parsed)
end

-- Build a Local signer from a base58 representation of the 64-byte
-- secret (the Phantom / Solflare export shape).
function M.base58(s)
  if type(s) ~= 'string' or #s == 0 then
    return nil, 'pay_kit: signer.base58: expected a non-empty string'
  end
  local base58 = require('pay_kit.solana.base58')
  local ok, bytes_or_err = pcall(base58.decode, s)
  if not ok then
    return nil, 'pay_kit: signer.base58: invalid base58: ' .. tostring(bytes_or_err)
  end
  if type(bytes_or_err) ~= 'string' or #bytes_or_err ~= 64 then
    return nil, 'pay_kit: signer.base58: decoded length must be 64 bytes'
  end
  return local_signer.new(bytes_or_err)
end

-- Build a Local signer from a 128-character hex string.
function M.hex(s)
  local bytes, err = hex_to_bytes(s)
  if err then return nil, err end
  return local_signer.new(bytes)
end

-- Read a Solana CLI keypair file (64-byte JSON array on disk).
function M.file(path)
  if type(path) ~= 'string' or path == '' then
    return nil, 'pay_kit: signer.file: expected a non-empty path'
  end
  local fh, open_err = io.open(path, 'rb')
  if not fh then
    return nil, 'pay_kit: signer.file: ' .. tostring(open_err)
  end
  local contents = fh:read('*a')
  fh:close()
  return M.json(contents)
end

-- Read an env var and auto-detect the encoding (JSON array, hex,
-- base58). Unset or empty env returns plain `nil` so the operator
-- option table's nil-as-no-opinion contract composes cleanly. Set
-- but malformed env returns `(nil, err)` because that is a real bug.
function M.from_env(name)
  if type(name) ~= 'string' or name == '' then
    return nil, 'pay_kit: signer.from_env: expected an env-var name'
  end
  local raw = os.getenv(name)
  if raw == nil or raw == '' then return nil end
  local trimmed = raw:match('^%s*(.-)%s*$')
  if trimmed == '' then return nil end
  if trimmed:sub(1, 1) == '[' then return M.json(trimmed) end
  if #trimmed == 128 and trimmed:match('^[0-9a-fA-F]+$') then
    return M.hex(trimmed)
  end
  return M.base58(trimmed)
end

-- Generate a fresh ephemeral keypair. Test-only; returns
-- `(nil, err)` when no keypair-generation backend is available
-- (the openssl-only environments cannot synthesise a Solana
-- 64-byte secret without seed derivation - production callers
-- load from files or env vars instead).
function M.generate()
  local ed25519 = require('pay_kit.util.ed25519')
  local secret, err = ed25519.generate()
  if not secret then return nil, err end
  return local_signer.new(secret)
end

return M
